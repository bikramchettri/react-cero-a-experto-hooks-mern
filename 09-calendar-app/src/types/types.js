export const types = {

    uiOpenModal: '[ui] Open modal',
    uiCloseModal: '[ui] Close modal',

    eventSetActive:'[event] Set active',
    eventAddNew:'[event] Add new',
    eventClearActive:'[event] Clear event active',
    eventUpdated:'[event] Event updated',
    eventDeleted:'[event] Event deleted',

    authChecking: '[auth] Checking login state',
    authCheckingFinish: '[auth] Finish checking login state',
    authStartLogin: '[auth] Start login',
    authLogin: '[auth] Checking Login',
    authStartRegister: '[auth] Start register',
    authStartTokenRenew: '[auth] Start token renew',
    authLogout: '[auth] Logout',

}