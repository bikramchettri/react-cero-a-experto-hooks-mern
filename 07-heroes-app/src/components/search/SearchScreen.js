import React from 'react';
import queryString from 'query-string';
import { heroes } from '../../data/heroes';
import { HeroCard } from '../heroes/HeroCard';
import { useForm } from '../../hooks/useForm';
import { useLocation} from 'react-router-dom';

export const SearchScreen = ({history}) => {

    const location  = useLocation();

    const { q = '' } = queryString.parse(location.search); //obtengo el query -> q
    const heroesFiltrados =  heroes;

    const [values, handleInputChange]= useForm({
        searchText: q //valor inicial
    });

    const {searchText} = values;

    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${searchText}`);

    }
    return (
        <div>
            <h1>Search Screen</h1>
            <hr/>
            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr/>
                    <form onSubmit={handleSearch}>
                        <input
                            type="text"
                            className="form-control"
                            name="searchText"
                            value={searchText}
                            onChange={handleInputChange}
                            autoComplete="off"
                            placeholder="Encuentra a tu heroe"
                        />
                        <button
                            type="submit"
                            className="btn btn-primary btn-block mt-1"
                        >
                        Search
                        </button>
                    </form>

                </div>
                <div className="col-7">
                    <h4>Resultados</h4>
                    <hr/>
                    {
                        heroesFiltrados.map( hero => (
                            <HeroCard 
                                key= {hero.id}
                                {...hero}
                            />
                        ))
                    }
                </div>
            </div>
        </div>
    )
}
