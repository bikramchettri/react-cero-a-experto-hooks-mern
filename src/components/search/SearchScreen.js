import React , { useMemo } from 'react';
import queryString from 'query-string'; //paquete de npm
import { HeroCard } from '../heroes/HeroCard';
import { useForm } from '../../hooks/useForm';
import { useLocation} from 'react-router-dom';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({history}) => {

    const location  = useLocation();

    const { q = '' } = queryString.parse(location.search); //obtengo el query -> q
    
    const [values, handleInputChange]= useForm({
        searchText: q //valor inicial
    });

    const {searchText} = values;
    //memorizo los valores en el query string
    const heroesFiltrados = useMemo(() => getHeroesByName(q), [q]); //si el query cambia disparo getHeroesByName
    
    //const heroesFiltrados =  getHeroesByName(searchText);

    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${searchText}`);

    }
    return (
        <div>
            <h1>Search Screen</h1>
            <hr/>
            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr/> 
                    <form onSubmit={handleSearch}>
                        <input
                            type="text"
                            className="form-control"
                            name="searchText"
                            value={searchText}
                            onChange={handleInputChange}
                            autoComplete="off"
                            placeholder="Encuentra a tu heroe"
                        />
                        <button
                            type="submit"
                            className="btn btn-primary btn-block mt-1"
                        >
                        Search
                        </button>
                    </form>

                </div>
                <div className="col-7">
                    <h4>Resultados</h4>
                    <hr/>
                    { (q === '') &&
                        <div className="alert alert-info">
                            Search a Hero
                        </div>
                    }

                    { (q !== '' && heroesFiltrados.length === 0) &&
                        <div className="alert alert-danger">
                            There is no a hero with { q }
                        </div>
                    }
                    
                    {
                        heroesFiltrados.map( hero => (
                            <HeroCard 
                                key= {hero.id}
                                {...hero}
                            />
                        ))
                    }
                </div>
            </div>
        </div>
    )
}
