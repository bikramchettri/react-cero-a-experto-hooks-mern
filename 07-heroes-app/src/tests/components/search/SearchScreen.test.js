import React  from 'react';
import {mount}  from 'enzyme';
import { MemoryRouter , Route} from 'react-router-dom';
import { types } from '../../../types/types';
import { SearchScreen } from '../../../components/search/SearchScreen';


describe('Pruebas en <SearchScreen/>', () => {

    
    test('Debe de mostrarse correctamente con valores por defecto', () => {
       
        const wrapper = mount(
           <MemoryRouter initialEntries={['/search']}>
               <Route path="/search" component= {SearchScreen}/>
           </MemoryRouter>
       )       

       expect(wrapper).toMatchSnapshot();        
       expect(wrapper.find('.alert-info').text().trim()).toBe('Search a Hero');        


    });

    test('Debe de mostrar a Batman y el input con el valor del queryString', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path="/search" component= {SearchScreen}/>
            </MemoryRouter>
        )
        
       expect(wrapper.find('input').prop('value')).toBe('batman');        
       expect(wrapper).toMatchSnapshot();        
        
    })
    
});