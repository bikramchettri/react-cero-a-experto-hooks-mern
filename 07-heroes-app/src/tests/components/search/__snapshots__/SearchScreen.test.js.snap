// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen/> Debe de mostrar a Batman y el input con el valor del queryString 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Encuentra a tu heroe"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-primary btn-block mt-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-mutted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                className="btn btn-primary"
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                MÃ¡s...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen/> Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Encuentra a tu heroe"
          type="text"
          value=""
        />
        <button
          className="btn btn-primary btn-block mt-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a Hero
      </div>
    </div>
  </div>
</div>
`;
